secp256kfun/src/poly.rs

```diff
@@ -0,0 +1,323 @@
+    indicies: &[Scalar<Public, impl ZeroChoice>],
+) -> Vec<Vec<Scalar<Public, Zero>>> {
+    // Calculated from the product of these indices coefficients:
+    //      l_j(x) = Product[ (x-x_m)/(x_j-x_m), j!=m ]
+    // Or
+    //      l_j(x) = Product[ a_m*x + b_m, j!=m], where a_m = 1/(x_j-x_m) and b_m = -x_m*a_m.
+    indicies
+        .iter()
+        .enumerate()
+        .map(|(j, x_j)| {
+            let mut coefficients: Vec<_> = vec![];
+            for (_, x_m) in indicies.iter().enumerate().filter(|(m, _)| *m != j) {
+                let a_m = s!(x_j - x_m)
+                    .non_zero()
+                    .expect("points must lie at unique indicies to interpolate")
+                    .invert();
+                let b_m = s!(-x_m * a_m).mark_zero();
+
+                // Multiply out the product. Beginning with the first two coefficients
+                // we then take the next set (b_1, a_1), multiply through, and collect terms.
+                if coefficients.is_empty() {
+                    coefficients.extend([b_m.mark_zero().public(), a_m.mark_zero().public()])
+                } else {
+                    let mut updated_coefficients = coefficients.clone();
+                    for i in 0..coefficients.len() {
+                        let bumping_up_degree = if i != 0 {
+                            s!({ &coefficients[i - 1] } * a_m)
+                        } else {
+                            s!(0)
+                        };
+                        let same_degree = s!({ &coefficients[i] } * b_m);
+                        updated_coefficients[i] = s!(same_degree + bumping_up_degree).public();
+                    }
+                    let higher_degree = s!({ coefficients.last().unwrap() } * a_m);
+                    updated_coefficients.push(higher_degree.public());
+                    coefficients = updated_coefficients;
+                }
+            }
+            coefficients
+        })
+        .collect()
+}
+
+/// Find the coefficients of the polynomial that interpolates a set of points (index, point).
+///
+/// Panics if the indicies are not unique.
+///
+/// A vector with a tail of zero coefficients means the interpolation was overdetermined.
+pub fn interpolate_point_polynomial(
+    points_at_indicies: Vec<(Scalar<Public, impl ZeroChoice>, Point)>,
+) -> Vec<Point<impl PointType, Public, Zero>> {
+    let (indicies, points): (Vec<_>, Vec<_>) = points_at_indicies.into_iter().unzip();
+
+    let basis_polynomials: Vec<_> = unit_basis_polys(indicies.as_slice());
+
+    let interpolating_basis: Vec<_> = basis_polynomials
+        .iter()
+        .zip(points)
+        .map(|(basis_poly, y_value)| point_mul(basis_poly, &y_value))
+        .collect();
+
+    interpolating_basis.into_iter().fold(
+        Vec::with_capacity(basis_polynomials[0].len()),
+        |acc, poly| add(&acc, &poly),
+    )
+}
+
+/// Multiply each coefficient in a scalar polynomial by a point.
+pub fn point_mul(
```

Comment:
I'd remove and inline this one. Perhaps I made a complaint that caused you to add it. I was probably confused that this was a single scalar against a point polynomial rather than the other way around. Multiplying a scalar vector by a point is fairly incoherent conceptually (it's usually the other way around!).